en: add doc about volume snapshot backup/restore across multiple kubernetes (#2406)
pingcap/docs-tidb-operator: en/deploy-br-federation-manager.md

Reviewer: Preview: https://preview-operator-pingcap-doc.pingcap-docs-website-scaffold-1hq.pages.dev/tidb-in-kubernetes/dev/deploy-br-federation-manager

Assume that you have 3 kubeconfig files with file paths: `kubeconfig-path1`, `kubeconfig-path2`, `kubeconfig-path3`. To merge these files into one kubeconfig file with file path `data-planes-kubeconfig`, execute the following command:

```shell
KUBECONFIG=${kubeconfig-path1}:${kubeconfig-path2}:${kubeconfig-path3} kubectl config view --flatten > ${data-planes-kubeconfig}
```

## Step 2: Deploy BR Federation Manager in the control plane

To deploy the BR Federation Manager, you need to select one Kubernetes cluster as the control plane. The following steps should **only be executed on the control plane**.

### Step 2.1: Create CRD

The BR Federation Manager uses [Custom Resource Definition (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) to extend Kubernetes. Before using the BR Federation Manager, you must create the CRD in your Kubernetes cluster. This operation only needs to be performed once.

```shell
kubectl create -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/federation-crd.yaml
```

> ### Step 2.2: Prepare kubeconfig secret

Reviewer 2: ```suggestion
### Step 2.2: Prepare the kubeconfig secret
```

Now that you already have a kubeconfig file of data planes, you need to encode the kubeconfig file into a secret. Take the following steps:

1. Encode the kubeconfig file:

    ```shell
    base64 -i ${kubeconfig-path}
    ```

2. Store the output from the previous step in a secret object.

> **Note**
>
> The name of the secret and the data key of the kubeconfig field must match the following example:
>
> ```yaml
> apiVersion: v1
> kind: Secret
> metadata:
>   name: br-federation-kubeconfig
> type: Opaque
> data:
>   kubeconfig: ${encoded-kubeconfig}
> ```

### Step 2.3: Install BR Federation Manager

This section describes how to install the BR Federation Manager using [Helm 3](https://helm.sh/docs/intro/install/).

- If you prefer to use the default configuration, follow the **Quick deployment** steps.

> - If you want to use a custom configuration, follow the **Custom deployment** steps.

Reviewer 2: Generally, please use consistent words for similar cases in documentation, while in blog posts, you can use different words to convey one meaning.
```suggestion
- If you prefer to use a custom configuration, follow the **Custom deployment** steps.
```

---
title: Deploy BR Federation Manager on Kubernetes
summary: Learn how to deploy BR Federation Manager on Kubernetes.

> aliases: ['/docs/tidb-in-kubernetes/dev/deploy-br-federation-manager/']

Reviewer: ```suggestion
```

---

# Deploy BR Federation Manager on Kubernetes

> 

Reviewer: Please explain what is BR Federation Manager. As this is a brand new term for our user documentation, many might wonder what it is and how it is different from BR.

This document describes how to deploy BR Federation Manager across multiple Kubernetes.

## Prerequisites

Before deploy BR Federation Manager on Kubernetes, you should satisfy conditions below:

* Kubernetes >= v1.12
* you have multiple Kubernetes clusters
* All the Kubernetes clusters that serve as data planes have deployed TiDB-Operator

> ## Deploy BR Federation Manager

Reviewer: ```suggestion
```

This heading feels somehow redundant and adds depth to the document structure.

You can start directly from `## Step 1: Generate Kubeconfig file in all Data Planes`.

> ### Generate Kubeconfig file in all Data Planes

Reviewer: ```suggestion
## Step 1: Generate Kubeconfig file in all Data Planes
```

BR federation manager manages Kubernetes clusters of data planes by accessing their API servers.
So it needs kubeconfig to authenticate and authorize in the API servers. The users or service accounts in the kubeconfig file need
have all the permissions of **backups.pingcap.com, restores.pingcap.com** CRD at least.

You can get the kubeconfig file from the Kubernetes cluster administrator. But we also provide a way to generate
the kubeconfig file by yourself if you have the permission to access all the data planes.

> #### Create Resources about RBAC in Data Plane

Reviewer: ```suggestion
### Step 1.1: Create Resources about RBAC in Data Plane
```

Reviewer: Numbering the steps will make it clearer for users to navigate through the doc. Can you please update all the other headings in this format?

We should create the resources below in every data plane to allow br federation manager to manipulate backup and restore CR.

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: br-federation-member
  namespace: tidb-admin
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: br-federation-manager:br-federation-member
rules:
- apiGroups:
  - pingcap.com
  resources:
  - backups
  - restores
  verbs:
  - '*'
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: br-federation-manager:br-federation-member
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: br-federation-manager:br-federation-member
subjects:
- kind: ServiceAccount
  name: br-federation-member
  namespace: tidb-admin
```

For Kubernetes with version >= 1.24, you should create service account secret manually for external applications to access the Kubernetes API server.

```yaml
apiVersion: v1
kind: Secret
type: kubernetes.io/service-account-token
metadata:
 name: br-federation-member-secret
 annotations:
   kubernetes.io/service-account.name: "br-federation-member"
```

> #### Generate Kubeconfig files

Reviewer: ```suggestion
### Step 1.2: Generate Kubeconfig files
```

Suppose that you have 3 kubeconfig files with file paths: `kubeconfig-path1`, `kubeconfig-path2`, `kubeconfig-path3`,
and you want to merge them to one kubeconfig file with file path `data-planes-kubeconfig`. You can execute the command below to merge kubeconfig files.

{{< copyable "shell-regular" >}}

```shell
KUBECONFIG=${kubeconfig-path1}:${kubeconfig-path2}:${kubeconfig-path3} kubectl config view --flatten > ${data-planes-kubeconfig}
```

### Deploy BR Federation Manager in Control Plane

You should select one Kubernetes cluster as control plane to deploy BR Federation Manager. You need only execute steps below in the control plane.

#### Create CRD

BR Federation Manager uses [Custom Resource Definition (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) to extend Kubernetes.
Therefore, to use BR Federation Manager, you must first create the CRD, which is a one-time job in your Kubernetes cluster.

{{< copyable "shell-regular" >}}

```shell
kubectl create -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/federation-crd.yaml
```

#### Prepare Kubeconfig Secret

> You already have a kubeconfig file of data planes. Now, you need encode the kubeconfig file to a secret.
> Firstly, encode the kubeconfig file by `base64 -i ${kubeconfig-path}`. Secondly, put the output of first step to a secret object.
> **Note, the name of the secret and the data key of kubeconfig field must be equal to which is in the example below**.

Reviewer: ```suggestion
You already have a kubeconfig file of data planes. Now, you need encode the kubeconfig file to a secret. Take the following steps:

1. Encode the kubeconfig file by `base64 -i ${kubeconfig-path}`.
2. Put the output of first step to a secret object.

> **Note:**
>
> The name of the secret and the data key of the kubeconfig field must be equal to what is in the example below.
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: br-federation-kubeconfig
type: Opaque
data:
  kubeconfig: ${encoded-kubeconfig}
```

#### Install BR Federation Manager

This section describes how to install BR Federation Manager using [Helm 3](https://helm.sh/docs/intro/install/).

> 

Reviewer: Can you please explain the difference between **quick deployment** and **custom deployment**?

* you have multiple Kubernetes clusters
* Every Kubernetes clusters can communicate with each other by network
* All the Kubernetes clusters that serve as data planes have deployed TiDB-Operator

## Deploy BR Federation Manager

### Generate Kubeconfig files in all Data Planes

BR federation manager manages Kubernetes clusters of data planes by accessing their API servers.
So it needs kubeconfing files to authenticate the API servers. The user or service account in the kubeconfig file need
have all the permissions of **backups.pingcap.com, restores.pingcap.com** CRD at least.

You can get kubeconfig files from the Kubernetes cluster administrator. But we also provide a way to generate
kubeconfig files by yourself if you have the permission to access all the data planes.

#### Create Resources about RBAC

We should create the resources below in every data plane to allow br federation manager to manipulate backup and restore CR.

{{< copyable "shell-regular" >}}

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: br-federation-member
  namespace: tidb-admin
---
apiVersion: rbac.authorization.k8s.io/v1

> kind: ClusterRole

Reviewer 3: qq: why do we need clusterRole here? can we use Role?

Author: You can use `Role`, if you only create backup/restore in one namespace. The reason why I use `ClusterRole` in the doc is that we aren't sure whether the user creates backup/restore in one namespace or in multiple namespaces. To avoid the error about permission, I write `ClusterRole` here.

export USER_TOKEN_VALUE=$(kubectl -n tidb-admin get secret/${USER_TOKEN_NAME} -o=go-template='{{.data.token}}' | base64 --decode)
export CURRENT_CONTEXT=$(kubectl config current-context)
export CURRENT_CLUSTER=$(kubectl config view --raw -o=go-template='{{range .contexts}}{{if eq .name "'''${CURRENT_CONTEXT}'''"}}{{ index .context "cluster" }}{{end}}{{end}}')
export CLUSTER_CA=$(kubectl config view --raw -o=go-template='{{range .clusters}}{{if eq .name "'''${CURRENT_CLUSTER}'''"}}"{{with index .cluster "certificate-authority-data" }}{{.}}{{end}}"{{ end }}{{ end }}')
export CLUSTER_SERVER=$(kubectl config view --raw -o=go-template='{{range .clusters}}{{if eq .name "'''${CURRENT_CLUSTER}'''"}}{{ .cluster.server }}{{end}}{{ end }}')

cat << EOF > {k8s-name}-kubeconfig
apiVersion: v1
kind: Config
current-context: ${CURRENT_CONTEXT}
contexts:
- name: ${CURRENT_CONTEXT}
  context:
    cluster: ${CURRENT_CONTEXT}
    user: br-federation-member
    namespace: kube-system
clusters:
- name: ${CURRENT_CONTEXT}
  cluster:
    certificate-authority-data: ${CLUSTER_CA}
    server: ${CLUSTER_SERVER}
users:
- name: br-federation-member
  user:

>     token: ${USER_TOKEN_VALUE}

Reviewer 3: Thanks @BornChanger, we are fine to use either token or certs. I think if we go with certs then we can ignore the token section right?

EOF
```

### Control Plane

You should select one Kubernetes cluster as control plane to deploy BR Federation Manager. You need only execute steps below in the control plane.

#### Create CRD

{{< copyable "shell-regular" >}}

```shell
kubectl create -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/federation-crd.yaml
```

#### Create Secret

You already have kubeconfig files of data planes. Now, you need encode all the kubeconfig files to a secret.

Firstly, encode every kubeconfig file by `base64 -i {k8s-name}-kubeconfig`.

Secondly, put all the output of first step to a secret object.

{{< copyable "shell-regular" >}}

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: br-federation-kubeconfig
type: Opaque

> data:

Reviewer 3: Can we use a single kube-config file, since kube-config file supports context, we can put multiple cluster in one kube-config file. Thanks~

Author: Yes, we can. I have created an [issue](https://github.com/pingcap/tidb-operator/issues/5183) to track it.

>   {k8s-name1}: {encoded-kubeconfig-1}

Reviewer 4: Instead of having a mapping from k8s-name to encoded kubeconfig, we can instead provide single kubeconfig that has multiple contexts, one for each of the k8s-name1, k8s-name2, etc.

Kubeconfigs already supports multiple contexts, so we don't need to invent a new format here map from clusters to kubeconfigs.
```
apiVersion: v1
kind: Config
clusters:
- cluster:
    server: https://k8s-name1
  name: k8s-name1
- cluster:
    server: https://k8s-name2
  name: k8s-name2
contexts:
- context:
    cluster:  k8s-name1
    user:  user
  name: k8s-name1
- context:
    cluster:  k8s-name2
    user:  user
  name: k8s-name2
preferences: {}
users:
- name:  user
  user:
    token: <secret-token-here>
```

From my brief review of the code, the main change is that https://github.com/pingcap/tidb-operator/blob/8b1aa3bcd3d3aacbfca04c31cbe7a9231758c2c7/pkg/controller/br_fed_config.go#L76 command line arg should instead point to a single kubeconfig with multiple contexts

Author: It's a good suggestion, we can change to use one kubeconfig and change the command line arg to a file path. But I think we still need a secret to store the kubeconfig so that we can mount the kubeconfig to the container of `tidb-operator` directly.

Author: I have created an [issue](https://github.com/pingcap/tidb-operator/issues/5183) to track it.

  {k8s-name2}: {encoded-kubeconfig-2}
  ...
```

#### Install BR Federation Manager

1. Create a namespace to create resources related to br federation manager.

{{< copyable "shell-regular" >}}

```shell

> kubectl create ns br-fed-admin

Reviewer 5: It's optional to create a new namespace.

Reviewer 3: we should be able to run br-fed-admin within the tidb namespace?

Reviewer 5: Yes.  We can.

```
2. Create the secret that contains all the encoded kubeconfig files in the namespace

{{< copyable "shell-regular" >}}

```shell
kubectl create -f {secret}.yaml -n br-fed-admin
```
3. Download tidb-operator repo and switch to release-1.5 branch. Enter the Helm chart directory `./charts/br-federation`
and install the Helm chart.

{{< copyable "shell-regular" >}}

```shell

> helm install --namespace br-fed-admin br-federation-manager . --set image=${br-federation-image}

Reviewer 6: wondering why we need to secret above for this? we got `MountVolume.SetUp failed for volume "federation-kubeconfig" : secret "br-federation-kubeconfig" not found` error during container creation.

Reviewer 3: I think if we get the whole manifest it should mounting this secrete to an environment var.

Author: The name of the secret can be configured by the `values.yaml`, the default configuration is [here](https://github.com/pingcap/tidb-operator/blob/release-1.5/charts/br-federation/values.yaml#L24)

---

> title: Deploy BR Federation Manager on Kubernetes

Author: This documentation should be split. I will add the documentations to `TOC.md` after split this documentation.
